preshed/bloom.pyx

```diff
@@ -54,40 +58,108 @@ cdef class BloomFilter:
 
 
 cdef bytes bloom_to_bytes(const BloomStruct* bloom):
-    py = array("L")
-    py.append(bloom.hcount)
-    py.append(bloom.length)
-    py.append(bloom.seed)
-    for i in range(bloom.length // sizeof(key_t)):
-        py.append(bloom.bitfield[i])
-    if hasattr(py, "tobytes"):
-        return py.tobytes()
-    else:
-        # Python 2 :(
-        return py.tostring()
+    cdef key_t pad = 0 # to differentiate new from old data format
+    cdef key_t version = 1 # hardcoded, can be incremented
+    prefix = struct.pack("<QQQQQ", pad, version, bloom.hcount, bloom.length, bloom.seed)
+    buflen = bloom.length // KEY_BITS
+    contents = [bloom.bitfield[i] for i in range(buflen)]
+    buffer = struct.pack(f"<{buflen}Q", *contents)
+    return prefix + buffer
 
 
 cdef void bloom_from_bytes(Pool mem, BloomStruct* bloom, bytes data):
-    py = array("L")
-    if hasattr(py, "frombytes"):
-        py.frombytes(data)
-    else:
-        py.fromstring(data)
-    bloom.hcount = py[0]
-    bloom.length = py[1]
-    bloom.seed = py[2]
-    bloom.bitfield = <key_t*>mem.alloc(bloom.length // sizeof(key_t), sizeof(key_t))
-    for i in range(bloom.length // sizeof(key_t)):
-        bloom.bitfield[i] = py[3+i]
+    # new-style memory structure (each unit is a key_t/64 bits):
+    # - pad1: 0 (not valid in old data)
+    # - ver: 1 (can be raised later if necessary
+    # (following values are same as old-style, except length semantics changed)
+    # - hcount: number of hashes
+    # - length: bitfield length in bits
+    # - seed: seed value for hashes
+
+    if len(data) < 40:
+        # unlikely but possible with old data
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    pad, ver, hcount, length, seed = struct.unpack("<QQQQQ", data[0:40])
+    if pad !=0:
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    assert ver == 1, "Unknown serialization version"
+
+    bloom.hcount = hcount
+    bloom.length = length
+    bloom.seed = seed
+    buflen = length // KEY_BITS
+    contents = struct.unpack(f"<{buflen}Q", data[40:])
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
```

Comment:
Whenever calling `mem.alloc`, ensure (either by checking or asserting) that the value is not `0` - cf https://github.com/explosion/spaCy/pull/4423 - just to avoid any nasty bugs down the line ;-)

Reply:
Thanks for the tip, I added asserts to check that there are no calls to alloc with 0 length. It doesn't seem to be the issue this time though...

Reply:
Ah that's unfortunate - so we'll need to keep digging ;-)